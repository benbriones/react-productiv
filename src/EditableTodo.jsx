import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";

/** Show editable todo item.
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove }) {

  const [isEditing, setIsEditing] = useState(false);
  const { id, title, description, priority } = todo;

  /** Toggle if this is being edited */
  function toggleEdit() {
    // setIsEditing(!isEditing);
    setIsEditing(isEditing => !isEditing);
  }

  /** Call remove fn passed to this. */
  function handleDelete() {
    remove(id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave(formData) {
    update(formData);
    toggleEdit();
  }

  return (
    <div className="EditableTodo">

      {isEditing
        ? <TodoForm initialFormData={todo} handleSave={handleSave} />
        : <div className="mb-3">
          <div className="float-end text-sm-end">
            <button
              className="EditableTodo-toggle btn-link btn btn-sm"
              onClick={toggleEdit}>
              Edit
            </button>
            <button
              className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
              onClick={handleDelete}>
              Del
            </button>
          </div>
          <Todo
            id={id}
            title={title}
            description={description}
            priority={priority}/>
        </div>
       }

    </div>
  );
}

export default EditableTodo;
