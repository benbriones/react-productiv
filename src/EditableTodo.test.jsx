import {describe, it, expect } from "vitest";

import React from "react";
import { fireEvent, render } from "@testing-library/react";
import EditableTodo from "./EditableTodo";

describe("EditableToDo", function () {
    const testTodo =
        { id:4, title: "celebrate bday", description: "for dinner", priority:1 }

  it("renders without crashing", function () {
    // const mockRemove = vi.spyOn(update)
    // const mockUpdate = vi.spyOn(update)
    render(<EditableTodo todo={testTodo}  />);
  });

  it("matches snapshot", function () {
    const { container } = render(<EditableTodo todo={testTodo} />)
    expect(container).toMatchSnapshot();
  });

  it ("rendered contains expected", function () {
    const { container } = render(<EditableTodo todo={testTodo} />)
    const topTodo = container.querySelector(".Todo")
    expect(topTodo).toContainHTML("celebrate");
  });

  it ("renders correct info and edit/delete button", function () {
    const { container } = render(<EditableTodo todo={testTodo} />)
    const topTodo = container.querySelector(".Todo")
    expect(topTodo).toContainHTML("celebrate bday");
    expect(container).toContainHTML("EditableTodo-delBtn");
    expect(container).toContainHTML("EditableTodo-toggle");

  });
  it ("renders correct info when isEditing is true", function () {
    const { container } = render(<EditableTodo todo={testTodo} />)
    const editBtn = container.querySelector(".EditableTodo-toggle");
    fireEvent.click(editBtn)
    expect(container).toContainHTML("NewTodoForm");
    expect(container).not.toContainHTML("EditableTodo-toggle");
    expect(container).not.toContainHTML("EditableTodo-delBt");

  });
});

