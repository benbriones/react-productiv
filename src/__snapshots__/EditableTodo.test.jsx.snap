// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditableToDo > matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
        id="4"
      >
        <div>
          <b>
            celebrate bday
          </b>
           
          <small>
            (Priority: 
            1
            )
          </small>
        </div>
        <div>
          <small>
            for dinner
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
